Improve McStas-NeXus roundtrip

g5t opened the pull request:
A number of instrument and component level `McStas` fields are not included in the current version of `eniius`'s `NXMcStas` serialisation to NeXus. Without this information, it would not be possible to reproduce an equivalent `McStas` instrument file from a generated NeXus file.

## Changes
This commit builds on top of #8 (sorry for the confluence) to implement writing hopefully-sufficient information into the NeXus structure such that the original `McStas`(`Script`) file/object can be recovered.

- Instrument level `PARAMETER`, `DEPENDENCY`, `DECLARE`, and `INITIALIZE` blocks are copied from the `McStasScript` object into the `NXinstrument` under a new (unvalidated) `NXcollection` group, `.../mcstas`.
- Component level `EXTEND`, `GROUP`, `JUMP`, `SPLIT`, and `WHEN` blocks are added directly to their associated component `NX*` object as `../mcstas_extend`, `.../mcstas_group`, `.../mcstas_jump`, `.../mcstas_split`, and `.../mstas_when`, respectively.
- Parameter handling is improved to include (C) `type`, `name`, `value`, and `comment` for `McStasScript` `DeclareVariable`s, and additionally `unit` for `libpyvinyl` `Parameter`s.
- Translation of Component information from `McStas` to NeXus now includes the `McStasScript` component instance, which provides access to more information including component-specific `WHEN` and `EXTEND` settings. This is probably most useful along with user-based patching of `McStasComp2NX` to replace or augment its component-type-named methods for supporting locally defined components or specialized instrument logic.
- The method of `NXMcStas` which combined consecutive affine transformations whenever possible has been moved to a stand-alone function since a) it was static and b) it might be necessary for components that _must_ redefine their position due to conceptual differences between `McStas` and NeXus (see, e.g., `Slit`).

## Side effects
Any pre-existing patching of `McStasComp2NX` will be broken by this commit.

## Possible improvements
The per-component `McStas` information could very easily be moved into single `NXcollection` instead of (up to) 6 independent entries. If you agree that this is worthwhile, I'll combine the existing `json.dumps(kwargs)` component-level `.../mcstas` entry with, `.../mcstas_component`, `.../mcstas_order/`, and the new entries into a single `NXcollection`.
This change would, presumably, require modifications elsewhere in `eniius`.

g5t committed:
[Fix] Issue 8, add __repr_ and __str_ for AffineRotate

g5t committed:
Replace libpyvinyl parameters by their value before serializing

g5t committed:
[Refactor] Round-trip improvements

- changed McStasComp2NX 'Component method' signature to incude the component instance
- refactored the initializer to simplify some logic and remove a try/except block
- pulled out _reduce_transforms since from McStas2NX since a) it is static and b) it could be used if a component needs to modify its placement
- changed NXMcStas to take a full McStasScript instrument object as input
- added McStas instrument parameters and configuration to NeXus structure at the NXinstrument level

- added from_mcstasscript class method for Eniius

g5t commented:
The test failed due to an extraneous semicolon in the `.instr` file at
https://github.com/mducle/eniius/blob/376a1a4543732d727b6d86b66805eae8d8a271f9/instruments/isis_mari.instr#L65
`McStas` syntax requires commas between instrument parameters, and no trailing separator. 
`McStasScript` apparently parses the `;` and then expects to find a final parameter specification which is not there. It then adds this missing parameter to its `self.declare_list` as an empty `libpyvinyl.Parameters.Parameter.Parameter`.

I'll push a commit shortly which has an improved translation method that is robust against empty parameters.
It will also re-introduce the numeric parsing of McStasScript parameters read from a .instr file, which I mistakenly removed.

g5t committed:
[Fix] missing parameter error

mducle commented on eniius/mcstas.py:
```diff
@@ -232,57 +232,68 @@ def getNXtype(cls, comp):
+    def default_translation(cls, comp, **kw):
+        nxtype = cls.getNXtype(comp)
+        nx2mcs = NX2COMP_MAP.get(nxtype, ({}, {}))[1]
+        return getattr(nexus, nxtype)(**{nx: getattr(comp, mcs) for nx, mcs in nx2mcs.items()})
```
This is a cool line of code!

g5t replied:
Hopefully it's not too convoluted :sweat_smile:

mducle commented on eniius/mcstas.py:
```diff
@@ -232,57 +232,68 @@ def getNXtype(cls, comp):
+            return f"{a} - {b}" if isinstance(a, str) or isinstance(b, str) else a - b
+
+        def neg_avg(a, b):
+            return f"-({a} + {b})/2" if isinstance(a, str) or isinstance(b, str) else (a + b) / 2
```
I don't understand why the string representation should be a negative average... and also, should the arithmetic representation also be negative - e.g. `-(a + b) / 2` or `(-a - b) / 2`?

In addition, is it likely that you get something like `a = -x` and `b = x` (in string where `x` is a variable name)? Should there be a check for this? (e.g. using regular expressions or something like `eval(f'({a} + {b})', None, {f'{a}':1.23})==0`?)

g5t replied:
Since the NeXus `NXslit` object is always symmetric about its reference point but the `McStas` `Slit` component may not be, if `a+b` is finite the `NXslit` needs to be translated to define the same opening.
Yesterday I thought the translation amount should be the negative value of the average 'edge' position -- but now I think I got that backwards.

If my current thinking is right, I'll replace `neg_avg` with
```
def avg(a, b):
    return f"({a} + {b})/2" if isinstance(a, str) or isinstance(b, str) else (a + b)/2
```

At the moment the extra translation is not added, in part because I was unsure if `NXtranslations` would accept a string-valued vector. The bigger question is, how to add even a numeric-valued extra translation?
- Should the component-named method optionally return an extra affine transformation?
- Should the method take the pre-defined transformation(s) group as input, and add the extra translation itself?

mducle replied:
Hmm... well a lot of the logic for dealing with transformations is not in the `McStasComp2NX` class where the method is, but is higher up in the `NXMcStas` class which calls `McStasComp2NX` - especially we would want to run `reduce_affine_transforms()` on the new concatenated list of previous transformations plus the new one from the component.

So it would probably make more sense to use the first option (the method return an optional extra transformation) but then to pass this back up to `NXMcStas`. E.g. we'd change [line 225](https://github.com/mducle/eniius/blob/8ea4888ebd061f51845ab3142e801e3b0ee26928/eniius/mcstas.py#L225) to:

```python
self.nxobj, self.extra_transform = method(mcstas_comp, **kwargs)
```

And then [line 588](https://github.com/mducle/eniius/blob/8ea4888ebd061f51845ab3142e801e3b0ee26928/eniius/mcstas.py#588) to something like:

```python
translated_component = McStasComp2NX(comp, order, self.NXtransformations(name), only_nx=only_nx, **mcpars)
if translated_component.extra_transform:   # Assuming the method returns None for extra_transforms if N/A
    self.affinelist[name] = reduce_affine_transforms(translated_component.extra_transform + self.affinelist[name])
    translated_component['transforms'] = self.NXtransformations(name)
return translated_component.nxobj
```

I think (!) that the extra transform should be applied last so the outer name for the `depends_on` is still ok... (not sure if I'm right though...)

g5t replied:
I've added a working implementation for this, which unfortunately requires the McStas instrument to contain the extra translation(s) in the component EXTEND block as `eniius_data`.
There's even a simple test to check that the whole thing works.

Getting it working in general without the extra `eniius_data` may not be feasible without parsing arbitrary C code to work-out the value of parameters.

mducle commented on eniius/mcstas.py:
```diff
@@ -494,13 +523,81 @@ def NXtransformations(self, name):
+def mcstasscript_parameter_name_or_value(parameter):
+    from mcstasscript.helper.mcstas_objects import DeclareVariable
+    from libpyvinyl.Parameters.Parameter import Parameter
+    from ast import literal_eval
```
This import of `literal_eval` doesn't seem to be used... (was it replaced by `_float_int_or_str`?)

g5t replied:
Good catch, I was trying out different ways to differentiate `int` and `float` from parsed `str` objects.
I'll remove the unused import.

g5t replied:
The import has been removed in the latest set of commits.

mducle commented on eniius/mcstas.py:
```diff
@@ -494,13 +523,81 @@ def NXtransformations(self, name):
+        
+
+def mcstasscript_parameter_name_or_value(parameter):
+    from mcstasscript.helper.mcstas_objects import DeclareVariable
+    from libpyvinyl.Parameters.Parameter import Parameter
```
Should these imports be global? (`eniius` has `mcstasscript` as a dependency and `mcstasscript` in turn depends on `libpyvinyl` so we can always be sure they will be importable).

g5t replied:
I think global imports are to-be-avoided in general, since they would get picked up by anyone that later uses `from eniius.mcstas import *`. 
The runtime import overhead difference between global and local scoping is _probably_ small. So I would suggest keeping these function-scoped.

mducle commented on eniius/mcstas.py:
```diff
@@ -494,13 +523,81 @@ def NXtransformations(self, name):
+
+def mcstasscript_instrument_to_nx(instrument):
+    # Follow the McStasScript instrument writing logic:
+    mcinst = NXcollection(name=instrument.name, version='3.3')
```
Can we get the McStas version from the McStasScript object rather than having it hard-coded?

g5t replied:
Only the major version at the moment, which might not be safe-enough since new features can be added in minor releases. Maybe I can convince @mads-bertelsen to change `McStas_instr.mccode_version` to be the full version number of the mcstas binary McStasScript is using?

mducle commented on eniius/mcstas.py:
```diff
@@ -568,6 +568,8 @@ def mcstasscript_parameter_to_nexus(parameter, index=None):
 
     attributes = {f: getattr(parameter, f) for f in fields if hasattr(parameter, f)}
+    # filter out empty-string (or anything equivalent to False) valued attributes since they cause problems :/
+    attributes = {k: v for k, v in attributes.items() if v}
```
Is this the fix for the empty parameter (resulting from an extraneous `;` in the `instr` file?)

g5t replied:
It must be, but I'm not sure if I didn't add another check!

The extraneous semicolon gives a parameter where all of the checked fields evaluate equivalent to `False` in the if statement. So `attributes` becomes and empty `dict` and `out = NXfield()`.

The slightly worrying-that-it-works bit is that the returned name is `parameter.name` which is `''`, and therefore that the line assigning the collection item becomes `nx[''] = NXfield()`. Does `nexusformat.nexus` do some magic to ignore empty names, or does this overwrite the whole `NXcollection` with an empy `NXfield` object?

g5t replied:
So, `nx[''] = NXfield()` adds an entry into the `NXcollection` with name `''` -- it doesn't look like the entry can be retrieved via the dictionary syntax (`nv['']` returns the `NXcollection`) but `Writer.to_json` transverses the `.items()` property, so it *is* included in output.

`conv_types` converts the empty `NXfield()` to a string-valued `'None'`, which works but doesn't seem right.
I'll add a check to avoid including the empty parameter into the collection entirely.

g5t replied:
I've added the check, plus a test case that checks JSON output to ensure no parameters named `''` are produced.

mducle reviewed (changes requested):
Looks good... just a few queries and maybe an extraneous import to be removed.

> The per-component McStas information could very easily be moved into single NXcollection instead of (up to) 6 independent entries. If you agree that this is worthwhile, I'll combine the existing json.dumps(kwargs) component-level .../mcstas entry with, .../mcstas_component, .../mcstas_order/, and the new entries into a single NXcollection.
This change would, presumably, require modifications elsewhere in eniius.

Yeah, I think it would require changes in `nexus.py` which writes out a McStas `instr` file from the NeXus data (it currently reads `mcstas_components` and `mcstas_order`). But this code would need to be updated to handle the extra fields in any case (I think that at the moment it ignores the extra fields but if you combine them into a single `NXcollection` then `mcstas_components` and `mcstas_order` will no longer exist at the top level and the tests will fail).

My preference would be to have a single `NXcollection` but we can leave that for a future PR if you don't have time.

So, it's up to you - if you want to finish the McStas -> NeXus -> McStas round trip then you'd need add changes to `nexus.py` in this PR, if you have time; otherwise let's leave it as separate fields and combine them in a future PR (or not as the case maybe!)

g5t committed:
[Add] relative depends_on link to affine transformation

g5t committed:
[Add] NeXus object bypass mechanism for JSON output

g5t committed:
[Add] missing dependency, [Fix] PR-raised issues

g5t committed:
[Fix] name shadowing

g5t committed:
[Fix] Empty-parameter [Add] test for output empty parameter

g5t committed:
[Add] missing changes intended for last commit

mducle reviewed (approved):
Looks good. Thanks!

Especially thanks for cleaning up my hack-ish parsing of the `eniius_data` block with some regexes... for adding some type-hinting (I really need to use this more regularly!) and for fixing issue #11.

Also, if we really need absolute paths for the `depends_on`, we can add that in the `NXMcStas` class in the `NXinstrument()` method...

g5t committed:
[Add] absolute depends_on path output

g5t committed:
[Refactor] eniius_data parsing, improve extra transform handling

- The parsing of eniius_data looked for any string between two double
  quote characters. This enforced the use of line continuation
  mid-string for very-long lines, which some IDEs give warnings about.
- To make those IDEs happier, now any string between = and ; following
  the phrase eniius_data is parsed. This allows C style string
  contatenation over multiple lines (each with two double quotes) without
  use of line continuation back slashes.

- In some cases, like the Slit component, there *must* be an extra
  transformation between the McStas position and the NeXus position due
  to different ideas about how the component can/will be used.
- In these cases the McStas file can now specify the extra
  transformations in the component EXTEND block as encoded JSON. These
  get decoded and converted to NeXus class objects which can then be
  appended to the component class transforms property, or even overwrite
  existing entries if one is clever, careful, or neither.
- To help ensure this functionality works, a very simple test instrument
  and case has been added in tests/

g5t committed:
[Move] test instrument so the CI can find it?

g5t committed:
[Add] include the updated test file

g5t committed:
[Move] test instr file back to test folder
